3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,225 @@
+          entries.forEach(entry => {
+            const creativeId = entry[0];
+            const extraData = entry[1];
+            dev().assert(!this.creativeMessageRouters_[creativeId],
+                'Duplicate new creative message for ' + creativeId);
+            this.creativeMessageRouters_[creativeId] =
+              new AmpAnalytics3pCreativeMessageRouter(
+                this.win_, this.iframeMessagingClient_, creativeId, extraData);
+          });
+        });
+    this.iframeMessagingClient_.registerCallback(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT,
+        messageContainer => {
+          let entries;
+          dev().assert(messageContainer.data &&
+              (entries = Object.entries(messageContainer.data)).length,
+              'Received empty events message');
+          entries.forEach(entry => {
+            const creativeId = entry[0];
+            const messages = entry[1];
+            try {
+              dev().assert(messages && messages.length,
+                  'Received empty events list for' + creativeId);
+              dev().assert(this.creativeMessageRouters_[creativeId],
+                  'Discarding event message received prior to new creative' +
+                  ' message for' + creativeId);
+              this.creativeMessageRouters_[creativeId]
+                  .sendMessagesToListener(messages);
+            } catch (e) {
+              dev().error(TAG_, 'Failed to send message to listener: ' +
+                  e.message);
+            }
+          });
+        });
+    this.iframeMessagingClient_.sendMessage(
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.READY);
+  }
+}
+
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative. Also sends response messages from the iframe meant for this
+ * particular creative.
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!Window} win The enclosing window object
+   * @param {!IframeMessagingClient} iframeMessagingClient Facilitates
+   * communication with the frame
+   * @param {!string} creativeId The ID of the creative to route messages
+   * to/from
+   * @param {string=} opt_extraData Extra data to be passed to the frame
+   */
+  constructor(win, iframeMessagingClient, creativeId, opt_extraData) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @private {!string} */
+    this.creativeId_ = creativeId;
+
+    /** @private {?string} */
+    this.extraData_ = opt_extraData;
+
+    /** @private {?function(!Array<!AmpAnalytics3pEvent>)} */
+    this.eventListener_ = null;
+
+    if (this.win_.onNewAmpAnalyticsInstance) {
+      try {
+        this.win_.onNewAmpAnalyticsInstance(this);
+      } catch (e) {
+        dev().error(TAG_, 'Exception thrown by onNewAmpAnalyticsInstance in' +
```

Comment:
Consider moving onNewAmpAnalyticsInstance check prior to creating AmpAnalytics3pCreativeMessageRouter as AmpAnalytics3pMessageRouter essentially expects it to work properly.  In other words, it would call router's sendMessagesToListener function even though client is not actually listening